Issue 27925 - Squiggle first line / entire parenthetical expression in remove unnecessary parentheses diagnostic

dibarbet opened the pull request:
See https://github.com/dotnet/roslyn/issues/27925

When a warning / error to remove unnecessary parentheses is created it will currently squiggle the first parentheses character and fade the open and closing parentheses.  This PR changes the functionality to squiggle (but not fade) the entire expression.  If the expression goes beyond one line, only the first line is squiggled.  The open and close parentheses remain faded.

CyrusNajmabadi commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -15,11 +17,16 @@ internal abstract class AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer<
         where TLanguageKindEnum : struct
         where TParenthesizedExpressionSyntax : SyntaxNode
     {
+        private readonly DiagnosticDescriptor _unnecessaryWithoutSuggestionWithoutFadeDescriptor;
```
this needs documentation.

CyrusNajmabadi commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -116,9 +123,31 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+                additionalLocations,
+                properties: null));
+
             context.ReportDiagnostic(Diagnostic.Create(
```
each diagnostic should get a comment explaining what it is for and what the distinction is here.

CyrusNajmabadi commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -116,9 +123,31 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+
+            if (expressionLineSpan.StartLinePosition.Line < expressionLineSpan.EndLinePosition.Line)
+            {
+                var lines = parenthesizedExpression.SyntaxTree.GetText().Lines;
```
doesn't GetText take a cancellation token?

CyrusNajmabadi replied:
this all needs commenting to explain what behavior is trying to be achieved.

CyrusNajmabadi commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -116,9 +123,31 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+                var spanUntilEndOfLine = TextSpan.FromBounds(parenthesizedExpressionLocation.SourceSpan.Start, expressionFirstLine.Span.End);
+                return Location.Create(parenthesizedExpression.SyntaxTree, spanUntilEndOfLine);
+            }
+            return parenthesizedExpressionLocation;
```
an easier way to express this is that you can just say that you want the span to go from the start of the parenthesizedExpr to the min of either it's endpoint, or the end of the line it's on.

dibarbet replied:
Thanks for the suggestion, that makes it much cleaner!

dibarbet commented on src/EditorFeatures/CSharpTest/RemoveUnnecessaryParentheses/RemoveUnnecessaryParenthesesTests.cs:
```diff
@@ -2280,5 +2287,86 @@ void Test(bool a)
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestUnnecessaryParenthesisDiagnosticSingleLineExpression()
+        {
+            var openParenthesesDiagnostic = GetRemoveUnnecessaryParenthesesDiagnostic("(", 4, 16, true);
```
Still investigating the best way to address this one.  I'm currently looking for a good way to handle nested markup spans (since there can be multiple diagnostics on the same character(s) but with different spans).

It seems possible with a little work to add/retrieve multiple spans from the document, but then the question is how to associate a particular span with the specific diagnostic that should be in the span.  If you have any ideas I'd be happy to hear them.

mavasani commented on src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs:
```diff
@@ -15,11 +19,20 @@ internal abstract class AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer<
                    new LocalizableResourceString(nameof(FeaturesResources.Remove_unnecessary_parentheses), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
                    new LocalizableResourceString(nameof(FeaturesResources.Parentheses_can_be_removed), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
         {
+            _unnecessaryWithoutSuggestionWithoutFadeDescriptor = CreateDescriptorWithId(
```
So, this descriptor does not seem to be passed to the base type and would not be part of the `SupportedDiagnostics`. It seems to work as the descriptor created in the base constructor [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/CodeStyle/AbstractCodeStyleDiagnosticAnalyzer.cs,51) seems to have the same ID. Might be better to instead use [this](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/CodeStyle/AbstractBuiltInCodeStyleDiagnosticAnalyzer.cs,18) overload of the base constructor, like we do [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs,85)?

dibarbet replied:
Thanks for the catch.  It makes sense to explicitly register it.

To use that constructor, I will additionally pull it down into [AbstractParenthesesDiagnosticAnalyzer](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/AbstractParenthesesDiagnosticAnalyzer.cs,12) since it is the parent of AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.  The other 2 diagnostic descriptors are created in [AbstractCodeStyleDiagnosticAnalyzer ](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/CodeStyle/AbstractCodeStyleDiagnosticAnalyzer.cs,18).  I'll need to duplicate those in AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer

Would it make sense to move the descriptor (and initialization) to [AbstractCodeStyleDiagnosticAnalyzer ](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/CodeStyle/AbstractCodeStyleDiagnosticAnalyzer.cs,18) as well, since it offers another option similar to the other two (e.g. no fade, no suggestion, but squiggle).
e.g. `UnnecessaryWithoutSuggestionWithoutFadeDescriptor = CreateDescriptorWithId...` inside the constructor of AbstractCodeStyleDiagnosticAnalyzer

@mavasani What do you think?

mavasani replied:
Sounds reasonable to me.

dibarbet commented:
Added a bunch of changes - while I was working on addressing Manish's feedback, I discovered an issue where when you hover over the parentheses character, 2 messages were shown.  This is because of the interaction between the isUnneccessaryFlag and Hidden diagnostics [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Diagnostics/DiagnosticsSquiggleTaggerProvider.cs,47).
However, it turns out if the message is empty then the message also will not show (from the same method).  

After fixing that issue, it led me down a rabbit whole of broken tests which I ended up needing to fix.
One issue was that a few tests ended up not working because it was not expecting two diagnostics to overlap (which is why I introduced the shouldVerifyEntireSpan flag), and other ones related to not expecting empty diagnostics.

mavasani commented on src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:
```diff
@@ -37,6 +37,7 @@ public struct TestParameters
             internal readonly int index;
             internal readonly CodeActionPriority? priority;
             internal readonly bool retainNonFixableDiagnostics;
+            internal readonly bool shouldVerifyEntireSpan;
```
>shouldVerifyEntireSpan [](start = 35, length = 22)

I think the flag name is misleading, especially given the default parameter value is false in the constructor, and might seem that we are skipping span verification by default. How about using an enum here:

```csharp
enum SpanVerificationKind
{
   Intersect,
   Match,
}
```

The default parameter value in constructor would be `SpanVerificationKind.Intersect`

mavasani commented on src/EditorFeatures/TestUtilities/Diagnostics/TestDiagnosticAnalyzerDriver.cs:
```diff
@@ -72,9 +72,19 @@ private async Task<IEnumerable<Diagnostic>> GetDiagnosticsAsync(DiagnosticAnalyz
             return allDiagnostics;
         }
 
+        private bool areSpansCompatible(TextSpan firstSpan, TextSpan secondSpan, bool shouldVerifyEntireSpan)
```
>areSpansCompatible [](start = 21, length = 18)

Nit: Either should be a local function OR the method name must start with a capital letter and must be a static method.

mavasani commented on src/EditorFeatures/TestUtilities/Diagnostics/TestDiagnosticAnalyzerDriver.cs:
```diff
@@ -72,9 +72,19 @@ private async Task<IEnumerable<Diagnostic>> GetDiagnosticsAsync(DiagnosticAnalyz
         public Task<IEnumerable<Diagnostic>> GetAllDiagnosticsAsync(DiagnosticAnalyzer workspaceAnalyzerOpt, Document document, TextSpan span)
         {
-            return GetDiagnosticsAsync(workspaceAnalyzerOpt, document, span, document.Project, getDocumentDiagnostics: true, getProjectDiagnostics: true);
+            return GetDiagnosticsAsync(workspaceAnalyzerOpt, document, span, document.Project, getDocumentDiagnostics: true, getProjectDiagnostics: true, false);
```
>false [](start = 154, length = 5)

We always use named argument specification for literal arguments, just like the prior arguments.

CyrusNajmabadi commented:
I would **really** prefer to not keep adding more test knobs to an already complex test harness. Can we just address this issue:

>One issue was that a few tests ended up not working because it was not expecting two diagnostics to overlap 

Why can't we just fix that assumption?  Or those tests?

mavasani commented:
@dibarbet Is it trivial to address Cyrus' concerns? Otherwise, I feel we have spent enough cycles on this to not warrant another round of onion peeling for the value add from this fix. I am fine with current change as is, or even going back to the commit where we showed duplicate messages in quick info. We can always come back to this if we hear customer feedback.

dibarbet commented:
> Why can't we just fix that assumption? Or those tests?

I'm not sure it would be extremely easy.  The assumption seems to be built into the code actions test runner.  The code actions test is relying on the fact that there is only one diagnostic returned, see http://source.roslyn.io/#Roslyn.Services.Test.Utilities/Diagnostics/AbstractUserDiagnosticTest.cs,196 where it will only get the fix(es) for the first diagnostic that is returned.  I think modifying this would require almost the same level of digging into the test runners as I've already done.

However I do have good news, I'm just blind and missed a similar knob that already exists.  TIL what the mysterious index parameter is used for.
So to answer 

> @dibarbet Is it trivial to address Cyrus' concerns?

Yes I can remove the knob I added because I can use a different one that already exists.
Apologies for the churn guys!

dibarbet commented:
Not quite ready for final review yet, fixing one small issue with the index tests.

dibarbet commented:
Ok I think it's good now.  To use the code actions index I had to modify it slightly to get code actions from more than just the first diagnostic.

Some more tests were assuming only one diagnostic so I fixed those up too.

dibarbet commented:
@jinujoseph ready for approval
